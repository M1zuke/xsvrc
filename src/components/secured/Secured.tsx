import React, { PropsWithChildren, ReactElement, useMemo } from 'react';
import { useSelector } from 'react-redux';
import { Login } from '../../container/views/login/Login';
import { savedCookies } from '../../store/cookies/selectors';
import { StoredCookie } from '../../store/cookies/state';
import { isErrorType, Loadable } from '../../store/reducer';
import { vrcUserInfo } from '../../store/user/selectors';
import { AuthenticatedUserInfo } from '../../store/user/state';

export function isLoggedIn(cookies: StoredCookie[], userInfo: Loadable<AuthenticatedUserInfo>): boolean {
  const cookie = isProbablyAuthenticated(cookies);
  const userInfoNull = userInfo === null;
  const userInfoIsErrorType = isErrorType(userInfo);
  return cookie && !userInfoNull && !userInfoIsErrorType;
}

export function isProbablyAuthenticated(cookies: StoredCookie[]): boolean {
  return !!cookies.find((cookie) => cookie.key === 'auth');
}

export function Secured({
  children,
}: PropsWithChildren<{
  /* TODO: Add Child Only Props */
}>): ReactElement {
  const cookies = useSelector(savedCookies);
  const userInfo = useSelector(vrcUserInfo);

  const isNotAuthenticated = useMemo(() => !isLoggedIn(cookies, userInfo), [cookies, userInfo]);

  if (isNotAuthenticated) {
    return <Login />;
  }

  return <>{children}</>;
}
