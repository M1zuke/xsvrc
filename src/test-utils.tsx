import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { createMemoryHistory, Location, MemoryHistory } from 'history';
import merge from 'lodash.merge';
import React, { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { Route, RouteProps, Router } from 'react-router-dom';
import { DeepPartial, Store } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { AppState } from './store';
import { AppAction, AppActions, AppActionsType } from './store/actions';
import { INITIAL_API_INFO_STATE } from './store/api-info/state';
import { INITIAL_FRIEND_INFO_STATE } from './store/friends/state';
import { INITIAL_HISTORY_STATE } from './store/history/state';
import { INITIAL_PERSISTED_STATE } from './store/persisted/state';
import { createAppStore } from './store/store';
import { INITIAL_USER_EVENT_STATE } from './store/user-events/state';
import { INITIAL_USER_STATE } from './store/user/state';
import { INITIAL_VIEW_STATE } from './store/view/state';
import { INITIAL_WORLD_STATE } from './store/worlds/state';

function Wrapper(
  props: React.PropsWithChildren<{
    store: Store;
    history: MemoryHistory;
    route: Pick<RouteProps, 'path'> & Partial<RouteProps>;
  }>,
): ReactElement {
  return (
    <Router history={props.history}>
      <Provider store={props.store}>
        <Route {...props.route}>{props.children}</Route>
      </Provider>
    </Router>
  );
}

export const INITIAL_TEST_APP_STATE: AppState = {
  persisted: INITIAL_PERSISTED_STATE,
  user: INITIAL_USER_STATE,
  view: INITIAL_VIEW_STATE,
  apiInfo: INITIAL_API_INFO_STATE,
  friends: INITIAL_FRIEND_INFO_STATE,
  userEvents: INITIAL_USER_EVENT_STATE,
  worlds: INITIAL_WORLD_STATE,
  history: INITIAL_HISTORY_STATE,
};

export function extendInitialState(state: DeepPartial<AppState>): AppState {
  return merge({}, INITIAL_TEST_APP_STATE, state);
}

export function createTestStore(state?: DeepPartial<AppState>): TestStore {
  return createAppStore(extendInitialState(state || {})) as TestStore;
}

export type LocationOverrides<S = Record<string, unknown>> =
  | DeepPartial<Omit<Location, 'state'>>
  | (DeepPartial<Omit<Location, 'state'>> & { state: S });

interface CustomRenderSettings<S> {
  options?: Omit<RenderOptions, 'wrapper'>;
  state?: DeepPartial<AppState>;
  location?: LocationOverrides<S>;
  route?: Partial<RouteProps>;
}

type FunctionPropertyNames<T> = { [K in keyof T]: T[K] extends (...args: never) => never ? K : never }[keyof T];
type SpyObject<T, F extends FunctionPropertyNames<T>> = Pick<T, Exclude<keyof T, F>> & {
  [P in F]: T[P] &
    jest.SpyInstance<
      ReturnType<T[P] extends (...args: never) => never ? T[P] : never>,
      Parameters<T[P] extends (...args: never) => never ? T[P] : never>
    >;
};

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export type HistorySpy = SpyObject<MemoryHistory<Record<string, unknown>>, 'push' | 'replace' | 'goBack'>;

export type TestStore = Store<AppState, AppActions> & {
  dispatch: ThunkDispatch<AppState, undefined, AppAction<AppActionsType>>;
};

export interface CustomRenderResult {
  result: RenderResult;
  store: TestStore;
  history: HistorySpy;
}

const initialLocation: Location<Record<string, unknown>> = {
  pathname: '/',
  search: '',
  state: {},
  hash: 'x122c123',
  key: '123',
};

// export function withAuthentication(state?: DeepPartial<AppState>): DeepPartial<AppState> {
//   return merge({}, state || {}, {
//     apiInfo: { clientApiKey: 'testApiKey' },
//     cookies: [{ key: 'auth', value: 'testAuthToken', url: 'testUrl' }],
//     userInfo: USER_INFO_FETCH_RESULT as AuthenticatedUserInfo,
//   });
// }

export function spyObject<T, F extends FunctionPropertyNames<T>>(object: T, functions: F[]): SpyObject<T, F> {
  Object.keys(object).forEach((prop) => {
    const field = prop as F;
    if (functions.includes(field)) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      //@ts-ignore
      jest.spyOn(object, field);
    }
  });

  return object as unknown as SpyObject<T, F>;
}

export function createTestHistory<S>(location?: LocationOverrides<S>): HistorySpy {
  const $location = merge({}, initialLocation, location) as Location<S>;
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const history = spyObject(createMemoryHistory(), ['push', 'replace', 'goBack']);
  history.replace($location); // move into test code?
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return history;
}

const defaultRoute: Pick<RouteProps, 'path'> = {
  path: '/',
};

function customRender<S = null>(ui: React.ReactElement, settings?: CustomRenderSettings<S>): CustomRenderResult {
  const store = createTestStore(settings?.state);
  const location = merge({}, initialLocation, settings?.location) as Location<S>;
  const history = createTestHistory<S>(location);

  const route = merge({}, defaultRoute, settings?.route);
  return {
    result: render(
      <Wrapper history={history} store={store} route={route}>
        {ui}
      </Wrapper>,
      settings?.options,
    ),
    store: store,
    history: history,
  };
}

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
