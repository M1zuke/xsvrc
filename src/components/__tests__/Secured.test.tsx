import pretty from 'pretty';
import React from 'react';
import { render, withAuthentication } from '../../test-utils';
import { Secured } from '../secured/Secured';

describe('<Secured /> Component', () => {
  it('renders Login if not logged in', () => {
    const customResult = render(
      <Secured>
        <div data-testid="test-content" />
      </Secured>,
    );
    expect(pretty(customResult.result.container.innerHTML)).toMatchSnapshot();
  });

  it('renders children if logged in', () => {
    const customResult = render(
      <Secured>
        <div data-testid="test-content" />
      </Secured>,
      {
        state: withAuthentication(),
      },
    );
    expect(pretty(customResult.result.container.innerHTML)).toMatchSnapshot();
  });
});
