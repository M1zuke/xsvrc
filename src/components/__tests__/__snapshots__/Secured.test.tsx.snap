// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Secured /> Component renders Login if not logged in 1`] = `
"<div class=\\"Component\\">
  <div class=\\"LoginBox\\">
    <div class=\\"LoginImage\\"></div><input class=\\"Component\\" type=\\"text\\" placeholder=\\"Username\\" value=\\"\\"><input class=\\"Component\\" type=\\"password\\" placeholder=\\"Password\\" value=\\"\\"><button class=\\"Component\\"><span class=\\"StyleElement\\">Login</span></button>
  </div>
</div>"
`;

exports[`<Secured /> Component renders children if logged in 1`] = `"<div data-testid=\\"test-content\\"></div>"`;
